### PR TITLE
MDN primary CDN

### PR DESCRIPTION
**Pull request by metadave**
@escattone here's a WIP for your Cloudfront work, it has not been applied (and would overwrite your test CDN). I'm fine with leaving this PR open while CDN behaviors get sorted out.

### CODE REVIEW

**Comment by escattone** [comment]
@metadave You're awesome, thanks so much, I can take it from here!

**Comment by metadave** [comment]
no need to update Terraform,  I can pull your changes that are made in the UI 😀

**Comment by escattone** [comment]
It gets even better, thanks again @metadave !

**Comment by metadave** [comment]
TODO: 

- [ ] sort out certificates
- [ ] create ALIAS records to point at stage|prod ELB, stop using [prod,stage].mdn.moz.works traffic policies

**Comment by jwhitlock** [comment]
We'll probably want to deploy the CDN in front of stage.mdn.moz.works a week or more before production. Is that possible with this PR, or does it need to be split?

**Comment by metadave** [comment]
@jwhitlock This PR includes `mdn-primary-cloudfrom-stage` and `mdn-primary-cloudfrom-prod` CDNs, using `stage.mdn.moz.works` and `prod.mdn.moz.works` respectively.

**Comment by escattone** [comment]
Here's my view on the steps required to place this CDN in front of...

**Stage**

- [ ] Ensure all of the CDN PR's (caching headers for all endpoints) have been deployed to stage.
- [ ] Create the new primary CDN (via this Terraform) aliased as `primary-stage.mdn.moz.works` with an origin of...hmm, it looks like we would go straight to the stage ELB `ad9c984f598bc11e7b1e70287e8e34e8-97307194.us-west-2.elb.amazonaws.com`?
- [ ] Make the DNS switch: `stage.mdn.moz.works` switched to the CDN alias `primary-stage.mdn.moz.works`. At this point, the static and media files will still be served from the CDN `stage-cdn.mdn.mozilla.net`.
- [ ] Stop using the `stage-cdn.mdn.mozilla.net` CDN for static files. Create, merge and deploy new https://github.com/mozmeao/infra PR with these new `stage.sh` (and `stage.mm.sh`) settings:
    ```
    export KUMA_MEDIA_URL=https://stage.mdn.moz.works/media/
    export KUMA_STATIC_URL=https://stage.mdn.moz.works/static/
    # also, should stage-cdn.mdn.mozilla.net be removed from `WEB_ALLOWED_HOSTS`?
    ```
- [ ] De-commission the `stage-cdn.mdn.mozilla.net` CDN.

**Production**

- [ ] Ensure all of the CDN PR's (caching headers for all endpoints) have been deployed to stage.
- [ ] Create the new primary CDN (via this Terraform) aliased as `primary-prod.mdn.moz.works` with an origin of `prod.mdn.moz.works`?
- [ ] Make the DNS switch: `developer.mozilla.org` switched to the CDN alias `primary-prod.mdn.moz.works`. At this point, the static and media files will still be served from the CDN `cdn.mdn.mozilla.net`.
- [ ] Stop using the `cdn.mdn.mozilla.net` CDN for static files. Create, merge and deploy new https://github.com/mozmeao/infra PR with these new `prod.sh` (and `prod.mm.sh`) settings:
    ```
    export KUMA_MEDIA_URL=https://stage.mdn.moz.works/media/
    export KUMA_STATIC_URL=https://stage.mdn.moz.works/static/
    # also, should cdn.mdn.mozilla.net be removed from `WEB_ALLOWED_HOSTS`?
    ```
- [ ] De-commission the `cdn.mdn.mozilla.net` CDN. 

How does this look to you all?

**Comment by jwhitlock** [comment]
We'll want to decom https://test-stage-cdn.mdn.moz.works/en-US/ as well, once stage is done.

**Comment by escattone** [comment]
@jwhitlock and talked in Paris about disabling both the GZip and Brotli middleware as a post-DNS-switch since the CDN will take care of GZip. This would improve the Kuma response time.

**Comment by escattone** [comment]
I added [this commit](https://github.com/mozmeao/infra/pull/726/commits/bab5a8499678561be431c068912a9d1c61e06e98) to ensure this PR reflects the two changes I made to the primary CDN for stage via the AWS console:
- increased the origin timeout from 30 sec to 60 sec (this was done to allow more time for slow endpoints like https://developer.allizom.org/en-US/dashboards/topic_lookup?topic=mathml when called with an "X-Requested-With: XMLHttpRequest" header)
- added a custom error response object with a min TTL of 0 sec for 504's so that slow responses from the origin that cause 504's are not cached

**Comment by metadave** [comment]
I will request a cert with the following names in `us-east-1` for use in Cloudfront:

- developer.mozilla.org
- mdn.mozillademos.org
- mdn-demos-origin.moz.works
- prod.mdn.moz.works
- developer.mozilla.com
- devmo.developer.mozilla.org
- mdn.mozilla.org
- developer-new.mozilla.org
- developers.mozilla.org

These are the names from the current developer.mozilla.org cert, do we need any changes? @escattone @jgmize @jwhitlock

**Comment by jwhitlock** [comment]
I think there are no changes, thanks @metadave

**Comment by metadave** [comment]
@jgmize cert requested via email verification

**Comment by jgmize** [comment]
Cert has been issued: arn:aws:acm:us-east-1:236517346949:certificate/144c40ad-1a60-4865-a252-58ff23961787

**Comment by metadave** [comment]
@jgmize can you take a look at this PR from a security perspective?

File: apps/mdn/mdn-aws/infra/mdn-cdn/cloudfront_primary/cloudfront.tf [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    smooth_streaming       = false
+    target_origin_id       = "${var.distribution_name}"
+    viewer_protocol_policy = "redirect-to-https"
+
+    forwarded_values {
+      headers = ["Host"]
+      query_string = false
+
+      cookies {
+        forward = "none"
```
Comment (jgmize) [inline apps/mdn/mdn-aws/infra/mdn-cdn/cloudfront_primary/cloudfront.tf]: It seems to me that this would prevent users from logging in.

**Comment by escattone** [comment]
I added another commit to reflect the changes made to the stage CDN via the AWS console, as well as add a `README.md` file to provide some context.

**Comment by escattone** [comment]
@metadave I'm not sure about one thing in the configuration, and that is how to specify that all headers should be forwarded within the `forwarded_values` section (see https://www.terraform.io/docs/providers/aws/r/cloudfront_distribution.html#headers). I used `headers = "*"`, but maybe it should be `headers = ["*"]` instead?

**Comment by metadave** [comment]
@escattone The docs are vague there, but `terraform plan` will fail and tell us what we should use. I'll check tomorrow.

**Comment by escattone** [comment]
Thanks @metadave! I amended my last commit to change from `headers = "*"` to `headers = ["*"]`.

**Comment by metadave** [comment]
- switched to cert arn specified in [this comment](https://github.com/mozmeao/infra/pull/726#issuecomment-379290507) 
- uncommented prod primary CDN section

**Comment by metadave** [comment]
sadly, Terraform [doesn't automatically set cache_behavior precedence](https://github.com/terraform-providers/terraform-provider-aws/issues/188), so we'll manually sort the 23 behaviors once each CDN is provisioned and leave a giant comment in the .tf

**Comment by stephendonner** [comment]
👍

**Comment by metadave** [comment]
Thanks for all the excellent work on this @escattone! 👏🎉

**Pull request merged**